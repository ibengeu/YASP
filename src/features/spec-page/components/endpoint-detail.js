"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useMemo, useState } from "react";
import { Badge } from "../../../core/components/ui/badge";
import { cn } from "../../../core/lib/utils";
import { ScrollArea } from "../../../core/components/ui/scroll-area";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "../../../core/components/ui/tabs";
import { SchemaTable } from "./schema-table";
const methodColors = {
    get: "bg-blue-500",
    post: "bg-green-500",
    put: "bg-orange-500",
    delete: "bg-red-500",
    patch: "bg-yellow-500",
    options: "bg-purple-500",
    head: "bg-cyan-500",
    trace: "bg-indigo-500",
};
export const EndpointDetail = ({ path, method, operation, components }) => {
    const [selectedContentType, setSelectedContentType] = useState(null);
    const [selectedResponseTypes, setSelectedResponseTypes] = useState({});
    // Use operation.requestBody directly, assuming it's already resolved
    const requestBody = useMemo(() => operation.requestBody, [operation.requestBody]);
    // Use operation.parameters directly, assuming they are already resolved
    const parameters = useMemo(() => (operation.parameters ?? []), [operation.parameters]);
    // Use operation.responses directly, assuming they are already resolved
    const responses = useMemo(() => (operation.responses ?? {}), [operation.responses]);
    const contentTypes = useMemo(() => (requestBody?.content ? Object.keys(requestBody.content) : []), [requestBody]);
    // Set default content type for request body
    useEffect(() => {
        if (contentTypes.length > 0 && selectedContentType === null) {
            setSelectedContentType(contentTypes[0]);
        }
    }, [contentTypes, selectedContentType]);
    // Set default content types for responses
    useEffect(() => {
        const newResponseTypes = {};
        Object.entries(responses).forEach(([code, response]) => {
            if (response.content) {
                const responseContentTypes = Object.keys(response.content);
                if (responseContentTypes.length > 0 && !selectedResponseTypes[code]) {
                    newResponseTypes[code] = responseContentTypes[0];
                }
            }
        });
        if (Object.keys(newResponseTypes).length > 0) {
            setSelectedResponseTypes(prev => ({ ...prev, ...newResponseTypes }));
        }
    }, [responses, selectedResponseTypes]);
    const handleResponseTypeChange = (code, contentType) => {
        setSelectedResponseTypes(prev => ({ ...prev, [code]: contentType }));
    };
    // Updated getSchemaType to assume schema is already resolved
    const getSchemaType = (schema) => {
        if (!schema)
            return "-";
        let resolvedSchema;
        if ("$ref" in schema) {
            const refKey = schema.$ref.split("/").pop();
            resolvedSchema = components?.schemas?.[refKey];
        }
        else {
            resolvedSchema = schema;
        }
        if (!resolvedSchema)
            return "-";
        if (resolvedSchema.type) {
            if (resolvedSchema.type === "array" && resolvedSchema.items) {
                const itemSchema = resolvedSchema.items;
                const itemType = getSchemaType(itemSchema);
                return `array of ${itemType}`;
            }
            return resolvedSchema.type;
        }
        if (resolvedSchema.properties) {
            return "object";
        }
        return "any";
    };
    return (_jsx("div", { className: "h-full overflow-hidden bg-stone-50/70", children: _jsx(ScrollArea, { className: "h-full", children: _jsxs("div", { className: "p-6", children: [_jsxs("div", { className: "flex items-center gap-4 mb-4", children: [_jsx(Badge, { className: cn("uppercase text-white", methodColors[method.toLowerCase()] ?? "bg-gray-500"), children: method }), _jsx("code", { className: "text-sm font-mono", children: path })] }), operation.summary && _jsx("h3", { className: "font-semibold mb-2", children: operation.summary }), operation.description && (_jsx("p", { className: "text-muted-foreground mb-6", children: operation.description })), parameters.length > 0 && (_jsxs("div", { className: "mb-6", children: [_jsx("h4", { className: "text-sm font-semibold mb-2", children: "Parameters" }), _jsx("div", { className: "rounded-lg border border-border shadow-sm bg-background", children: parameters.map((parameter, index) => (_jsxs("div", { className: cn("flex flex-col py-3 px-4 border-b border-border/50 last:border-0", "hover:bg-muted/50 transition-colors"), children: [_jsxs("div", { className: "flex items-center gap-3", children: [_jsxs("div", { className: "flex-1 min-w-0", children: [_jsxs("div", { className: "flex items-center gap-2", children: [_jsx("span", { className: "font-mono text-sm font-medium text-foreground truncate", children: parameter.name }), _jsx(Badge, { variant: "outline", className: "text-xs bg-slate-100 text-slate-700", children: parameter.in })] }), _jsx("div", { className: "text-sm text-muted-foreground", children: getSchemaType(parameter.schema) })] }), _jsx(Badge, { variant: parameter.required ? "destructive" : "secondary", className: cn("text-xs font-medium", parameter.required
                                                        ? "bg-red-100 text-red-700"
                                                        : "bg-slate-100 text-slate-700"), children: parameter.required ? "Required" : "Optional" })] }), parameter.description && (_jsx("div", { className: "mt-1 text-sm text-muted-foreground pl-2", children: parameter.description }))] }, index))) })] })), requestBody?.content && selectedContentType && requestBody.content[selectedContentType]?.schema && (_jsxs("div", { className: "mb-6", children: [_jsx("h4", { className: "text-sm font-semibold mb-2", children: "Request Body" }), _jsx("div", { className: "mb-4", children: _jsx(SchemaTable, { schema: requestBody.content[selectedContentType].schema, components: components }) })] })), Object.keys(responses).length > 0 && (_jsxs("div", { className: "space-y-6", children: [_jsx("h4", { className: "text-sm font-semibold mb-2", children: "Responses" }), Object.entries(responses).map(([code, response]) => {
                                if (!response) {
                                    return (_jsxs("div", { className: "mb-4", children: [_jsx(Badge, { variant: "outline", children: code }), _jsx("span", { className: "text-sm text-muted-foreground ml-2", children: "No response details" })] }, code));
                                }
                                const responseContentTypes = response.content ? Object.keys(response.content) : [];
                                const selectedResponseType = selectedResponseTypes[code] ?? responseContentTypes[0];
                                return (_jsxs("div", { className: "mb-6 rounded-lg", children: [_jsxs("div", { className: "flex items-center gap-2 mb-4", children: [_jsx(Badge, { variant: code.startsWith("2")
                                                        ? "default"
                                                        : code.startsWith("4") || code.startsWith("5")
                                                            ? "destructive"
                                                            : "outline", className: "text-sm", children: code }), _jsx("span", { className: "text-sm font-medium", children: response.description ?? "No description" })] }), response.content && responseContentTypes.length > 1 && (_jsx("div", { className: "mb-4", children: _jsxs(Tabs, { value: selectedResponseType, onValueChange: value => handleResponseTypeChange(code, value), className: "w-full", children: [_jsx(TabsList, { className: "mb-2 bg-stone-100/70", children: responseContentTypes.map(type => (_jsx(TabsTrigger, { value: type, className: "text-xs", children: type }, type))) }), responseContentTypes.map(contentType => (_jsx(TabsContent, { value: contentType, children: response.content && response.content[contentType]?.schema && (_jsx(SchemaTable, { schema: response.content[contentType].schema, components: components })) }, contentType)))] }) })), response.content &&
                                            responseContentTypes.length === 1 &&
                                            selectedResponseType && (_jsxs("div", { className: "mt-2", children: [_jsx(Badge, { variant: "outline", className: "mb-2", children: selectedResponseType }), response.content[selectedResponseType]?.schema && (_jsx(SchemaTable, { schema: response.content[selectedResponseType].schema, components: components }))] }))] }, code));
                            })] }))] }) }) }));
};
