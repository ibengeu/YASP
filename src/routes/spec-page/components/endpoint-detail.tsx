"use client"

import React, {useEffect, useMemo, useState} from "react"
import {Badge} from "@/components/ui/badge"
import {cn} from "@/lib/utils"
import {ScrollArea} from "@/components/ui/scroll-area"
import type {
    ComponentsObject,
    OperationObject,
    ParameterObject,
    ReferenceObject,
    ResponseObject,
    SchemaObject
} from "@/common/swagger.types.ts"
import {SchemaTable} from "@/routes/spec-page/components/schema-table"
import {Tabs, TabsContent, TabsList, TabsTrigger} from "@/components/ui/tabs"

// Type guard to check if an object is a ReferenceObject
function isReferenceObject(obj: unknown): obj is ReferenceObject {
    return obj !== null && typeof obj === 'object' && '$ref' in (obj as Record<string, unknown>);
}

interface EndpointDetailProps {
    path: string
    method: string
    operation: OperationObject
    components: ComponentsObject
}

const methodColors: Record<string, string> = {
    get: "bg-blue-500",
    post: "bg-green-500",
    put: "bg-orange-500",
    delete: "bg-red-500",
    patch: "bg-yellow-500",
    options: "bg-purple-500",
    head: "bg-cyan-500",
    trace: "bg-indigo-500",
}

export const EndpointDetail: React.FC<EndpointDetailProps> = ({path, method, operation, components}) => {
    const [selectedContentType, setSelectedContentType] = useState<string | null>(null)

    const [selectedResponseTypes, setSelectedResponseTypes] = useState<Record<string, string>>({})

    const requestBody = operation.requestBody && !isReferenceObject(operation.requestBody)
        ? operation.requestBody
        : null;

    const contentTypes = useMemo(() =>
            requestBody?.content ? Object.keys(requestBody.content) : [],
        [requestBody]
    );

    useEffect(() => {
        if (contentTypes.length > 0 && !selectedContentType) {
            setSelectedContentType(contentTypes[0])
        }
    }, [contentTypes, selectedContentType])

    useEffect(() => {
        if (operation.responses) {
            Object.entries(operation.responses).forEach(([code, response]) => {
                if (!isReferenceObject(response)) {
                    const typedResponse = response as ResponseObject
                    if (typedResponse.content && Object.keys(typedResponse.content).length > 0) {
                        const responseContentTypes = Object.keys(typedResponse.content)
                        if (responseContentTypes.length > 0 && !selectedResponseTypes[code]) {
                            setSelectedResponseTypes((prev) => ({
                                ...prev,
                                [code]: responseContentTypes[0],
                            }))
                        }
                    }
                }
            })
        }
    }, [operation.responses, selectedResponseTypes])

    const handleResponseTypeChange = (code: string, contentType: string) => {
        setSelectedResponseTypes((prev) => ({
            ...prev,
            [code]: contentType,
        }))
    }

    const getSchemaType = (schema: SchemaObject | ReferenceObject | undefined): string => {
        if (!schema) return "-";
        if (isReferenceObject(schema)) return "reference";
        return schema.type || "-";
    }

    const parameters = operation.parameters
            ?.filter((param): param is ParameterObject => !isReferenceObject(param))
        || [];

    return (
        <div className="h-full overflow-hidden">
            <ScrollArea className="h-full">
                <div className="p-6">
                    <div className="flex items-center gap-4 mb-4">
                        <Badge
                            className={cn("uppercase text-white", methodColors[method.toLowerCase()] || "bg-gray-500")}>{method}</Badge>
                        <code className="text-sm font-mono">{path}</code>
                    </div>

                    {operation.summary && <h3 className="font-semibold mb-2">{operation.summary}</h3>}

                    {operation.description && <p className="text-muted-foreground mb-6">{operation.description}</p>}

                    {parameters.length > 0 && (
                        <div className="mb-6">
                            <h4 className="text-sm font-semibold mb-2">Parameters</h4>
                            <div className="rounded-lg border overflow-hidden">
                                <table className="w-full border-collapse">
                                    <thead>
                                    <tr className="bg-stone-100/70 border-b">
                                        <th className="px-4 py-2 text-left font-semibold">Name</th>
                                        <th className="px-4 py-2 text-left font-semibold">In</th>
                                        <th className="px-4 py-2 text-left font-semibold">Type</th>
                                        <th className="px-4 py-2 text-left font-semibold">Required</th>
                                        <th className="px-4 py-2 text-left font-semibold">Description</th>
                                    </tr>
                                    </thead>
                                    <tbody>
                                    {parameters.map((parameter, index) => (
                                        <tr key={index} className="border-b last:border-0">
                                            <td className="px-4 py-2 font-mono">{parameter.name}</td>
                                            <td className="px-4 py-2">{parameter.in}</td>
                                            <td className="px-4 py-2">{getSchemaType(parameter.schema)}</td>
                                            <td className="px-4 py-2">
                                                <Badge variant={parameter.required ? "destructive" : "secondary"}>
                                                    {parameter.required ? "Required" : "Optional"}
                                                </Badge>
                                            </td>
                                            <td className="px-4 py-2 text-muted-foreground">{parameter.description || "-"}</td>
                                        </tr>
                                    ))}
                                    </tbody>
                                </table>
                            </div>
                        </div>
                    )}

                    {requestBody?.content && (
                        <div className="mb-6">

                            {selectedContentType && requestBody.content[selectedContentType] && (
                                <>
                                    <h4 className="text-sm font-semibold mb-2">Request Body</h4>
                                    <div className="mb-4">
                                        <div className="flex items-center gap-2 mb-2">
                                            <Badge variant="outline">{selectedContentType}</Badge>
                                            {requestBody.required &&
                                                <Badge variant="destructive">Required</Badge>}
                                        </div>
                                        <SchemaTable
                                            schema={requestBody.content[selectedContentType].schema as SchemaObject | ReferenceObject}
                                            components={components}
                                        />
                                    </div>
                                </>
                            )}
                        </div>
                    )}

                    {operation.responses && (
                        <div className="space-y-6">
                            <h4 className="text-sm font-semibold mb-2">Responses</h4>
                            {Object.entries(operation.responses).map(([code, response]) => {
                                if (isReferenceObject(response)) {
                                    return (
                                        <div key={code} className="mb-4">
                                            <div className="flex items-center gap-2 mb-2">
                                                <Badge
                                                    variant={
                                                        code.startsWith("2")
                                                            ? "default"
                                                            : code.startsWith("4") || code.startsWith("5")
                                                                ? "destructive"
                                                                : "outline"
                                                    }
                                                >
                                                    {code}
                                                </Badge>
                                                <span
                                                    className="text-sm text-muted-foreground">Reference: {response.$ref}</span>
                                            </div>
                                        </div>
                                    );
                                }

                                const typedResponse = response as ResponseObject;

                                if (!typedResponse.content || Object.keys(typedResponse.content).length === 0) {
                                    return (
                                        <div key={code} className="mb-4">
                                            <div className="flex items-center gap-2 mb-2">
                                                <Badge
                                                    variant={
                                                        code.startsWith("2")
                                                            ? "default"
                                                            : code.startsWith("4") || code.startsWith("5")
                                                                ? "destructive"
                                                                : "outline"
                                                    }
                                                >
                                                    {code}
                                                </Badge>
                                                <span
                                                    className="text-sm text-muted-foreground">{typedResponse.description}</span>
                                            </div>
                                        </div>
                                    )
                                }

                                const responseContentTypes = Object.keys(typedResponse.content || {})
                                const selectedResponseType = selectedResponseTypes[code] || responseContentTypes[0]

                                return (
                                    <div key={code} className="mb-6 border rounded-lg p-4 bg-stone-50/70">
                                        <div className="flex items-center gap-2 mb-4">
                                            <Badge
                                                variant={
                                                    code.startsWith("2")
                                                        ? "default"
                                                        : code.startsWith("4") || code.startsWith("5")
                                                            ? "destructive"
                                                            : "outline"
                                                }
                                                className="text-sm"
                                            >
                                                {code}
                                            </Badge>
                                            <span className="text-sm font-medium">{typedResponse.description}</span>
                                        </div>

                                        {typedResponse.content && responseContentTypes.length > 1 && (
                                            <div className="mb-4">
                                                <Tabs
                                                    value={selectedResponseType}
                                                    onValueChange={(value) => handleResponseTypeChange(code, value)}
                                                    className="w-full"
                                                >
                                                    <TabsList className="mb-2 bg-stone-100/70">
                                                        {responseContentTypes.map((type) => (
                                                            <TabsTrigger key={type} value={type} className="text-xs">
                                                                {type}
                                                            </TabsTrigger>
                                                        ))}
                                                    </TabsList>

                                                    {responseContentTypes.map((contentType) => (
                                                        <TabsContent key={contentType} value={contentType}>
                                                            {typedResponse.content && typedResponse.content[contentType] && typedResponse.content[contentType].schema && (
                                                                <SchemaTable
                                                                    schema={typedResponse.content[contentType].schema as SchemaObject | ReferenceObject}
                                                                    components={components}
                                                                />
                                                            )}
                                                        </TabsContent>
                                                    ))}
                                                </Tabs>
                                            </div>
                                        )}

                                        {typedResponse.content && responseContentTypes.length === 1 && selectedResponseType && (
                                            <div className="mt-2">
                                                <Badge variant="outline" className="mb-2">
                                                    {selectedResponseType}
                                                </Badge>
                                                {typedResponse.content[selectedResponseType] && typedResponse.content[selectedResponseType].schema && (
                                                    <SchemaTable
                                                        schema={typedResponse.content[selectedResponseType].schema as SchemaObject | ReferenceObject}
                                                        components={components}
                                                    />
                                                )}
                                            </div>
                                        )}
                                    </div>
                                )
                            })}
                        </div>
                    )}
                </div>
            </ScrollArea>
        </div>
    )
}