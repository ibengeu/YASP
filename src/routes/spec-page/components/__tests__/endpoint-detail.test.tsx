import {render, screen, within} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {describe, expect, it, vi} from 'vitest';
import type {ComponentsObject, OperationObject} from '@/common/swagger.types';
import '@testing-library/jest-dom';
import React, {ReactNode} from "react";
import {EndpointDetail} from "@/routes/spec-page/components/endpoint-detail.tsx";

vi.mock('@/components/ui/badge', () => ({
    Badge: ({children, variant, className}: { children: React.ReactNode; variant?: string; className?: string }) => (
        <span data-testid="badge" className={className} data-variant={variant}>
      {children}
    </span>
    ),
}));
vi.mock('@/components/ui/scroll-area', () => ({
    ScrollArea: ({children}: { children: React.ReactNode }) => <div>{children}</div>,
}));
vi.mock('@/components/ui/tabs', () => ({
    Tabs: ({children, value, onValueChange}: {
        children: React.ReactNode;
        value: string;
        onValueChange: (value: string) => void
    }) => (
        <div data-testid="tabs" data-value={value}
             onClick={(e) => onValueChange((e.target as HTMLElement).dataset.value || '')}>
            {children}
        </div>
    ),
    TabsList: ({children}: { children: React.ReactNode }) => <div data-testid="tabs-list">{children}</div>,
    TabsTrigger: ({children, value}: { children: React.ReactNode; value: string }) => (
        <button data-testid="tabs-trigger" data-value={value}>
            {children}
        </button>
    ),
    TabsContent: ({children, value}: { children: React.ReactNode; value: string }) => (
        <div data-testid="tabs-content" data-value={value}>
            {children}
        </div>
    ),
}));
vi.mock('@/routes/spec-page/components/schema-table', () => ({
    SchemaTable: ({schema}: { schema: unknown }) => <div data-testid="schema-table">{JSON.stringify(schema)}</div>,
}));
vi.mock('@/lib/utils', () => ({
    cn: (...args: string[]) => args.join(' '),
}));
const defaultProps = {
    path: '/users',
    method: 'get',
    operation: {summary: 'Get all users'} as OperationObject,
    components: {} as ComponentsObject,
};

function setup(jsx: ReactNode) {
    return {
        user: userEvent.setup(),
        ...render(jsx),
    }
}

describe('EndpointDetail', () => {
    it('displays HTTP method, path, and summary', () => {
        render(<EndpointDetail {...defaultProps} />);
        expect(screen.getByTestId('badge')).toHaveTextContent(/GET/i);
        expect(screen.getByText('/users')).toBeInTheDocument();
        expect(screen.getByText('Get all users')).toBeInTheDocument();
    });
    it('displays description when provided', () => {
        const props = {
            ...defaultProps,
            method: 'post',
            operation: {summary: 'Create a user', description: 'Creates a new user in the system.'},
        };
        setup(<EndpointDetail {...props} />);
        expect(screen.getByText('Creates a new user in the system.')).toBeInTheDocument();
    });
    it('displays parameters table when parameters are provided', () => {
        const props = {
            ...defaultProps,
            path: '/users/{id}',
            operation: {
                parameters: [
                    {name: 'id', in: 'path', required: true, description: 'User ID', schema: {type: 'string'}},
                    {name: 'limit', in: 'query', required: false, schema: {type: 'integer'}},
                ],
            },
        };
        setup(<EndpointDetail {...props} />);
        const table = screen.getByRole('table');
        const rows = within(table).getAllByRole('row');
        expect(rows).toHaveLength(3);
        expect(within(rows[1]).getByText('id')).toBeInTheDocument();
        expect(within(rows[1]).getByText('path')).toBeInTheDocument();
        expect(within(rows[1]).getByText('string')).toBeInTheDocument();
        expect(within(rows[1]).getByTestId('badge')).toHaveTextContent('Required');
        expect(within(rows[1]).getByText('User ID')).toBeInTheDocument();
        expect(within(rows[2]).getByText('limit')).toBeInTheDocument();
        expect(within(rows[2]).getByText('query')).toBeInTheDocument();
        expect(within(rows[2]).getByText('integer')).toBeInTheDocument();
        expect(within(rows[2]).getByTestId('badge')).toHaveTextContent('Optional');
        expect(within(rows[2]).getByText('-')).toBeInTheDocument();
    });
    it('displays request body with content type', async () => {
        const props = {
            ...defaultProps,
            method: 'post',
            operation: {
                requestBody: {
                    required: true,
                    content: {
                        'application/json': {schema: {type: 'object', properties: {name: {type: 'string'}}}},
                        'multipart/form-data': {
                            schema: {type: 'object', properties: {file: {type: 'string', format: 'binary'}}},
                        },
                    },
                },
            },
        };
        setup(<EndpointDetail {...props} />);
        expect(screen.getByText('Request Body')).toBeInTheDocument();
        const requestBodySection = screen.getByText('Request Body').closest('div')!;
        const contentTypeBadge = within(requestBodySection).getByText(/application\/json/);
        expect(contentTypeBadge).toBeInTheDocument();
        const requiredBadge = within(requestBodySection).getByText(/Required/);
        expect(requiredBadge).toBeInTheDocument();
        expect(screen.getByTestId('schema-table')).toHaveTextContent('name');
    });
    it('displays responses with content type selection', async () => {
        const props = {
            ...defaultProps,
            operation: {
                responses: {
                    '200': {
                        description: 'Successful response',
                        content: {
                            'application/json': {schema: {type: 'array', items: {type: 'string'}}},
                            'text/plain': {schema: {type: 'string'}},
                        },
                    },
                    '404': {description: 'Not found'},
                },
            },
        };
        setup(<EndpointDetail {...props} />);
        expect(screen.getByText('Responses')).toBeInTheDocument();
        const responsesSection = screen.getByText('Responses').closest('div')!;
        const status200Section = within(responsesSection).getByText('Successful response').closest('div')!;
        expect(within(status200Section).getByTestId('badge')).toHaveTextContent(/200/);
        expect(screen.getByText('Successful response')).toBeInTheDocument();
        expect(within(responsesSection).getByTestId('tabs')).toBeInTheDocument();
        const tabs = within(responsesSection).getByTestId('tabs');
        const jsonTrigger = within(tabs).getByText(/application\/json/);
        const textTrigger = within(tabs).getByText(/text\/plain/i);
        expect(jsonTrigger).toBeInTheDocument();
        expect(textTrigger).toBeInTheDocument();
        const schemaTables = within(tabs).getAllByTestId('schema-table');
        expect(schemaTables[0]).toHaveTextContent('array');
        await userEvent.click(textTrigger);
        await screen.findByText('text/plain');
        expect(schemaTables[1]).toHaveTextContent('string');
        const status404Section = within(responsesSection).getByText('Not found').closest('div')!;
        expect(within(status404Section).getByTestId('badge')).toHaveTextContent(/404/);
        expect(screen.getByText('Not found')).toBeInTheDocument();
    });
    it('handles reference objects correctly', () => {
        const props = {
            ...defaultProps,
            method: 'post',
            operation: {
                parameters: [{$ref: '#/components/parameters/UserId'}],
                requestBody: {$ref: '#/components/requestBodies/User'},
                responses: {'200': {$ref: '#/components/responses/Success'}},
            },
        };
        setup(<EndpointDetail {...props} />);
        expect(screen.queryByRole('table')).not.toBeInTheDocument();
        expect(screen.queryByText('Request Body')).not.toBeInTheDocument();
        expect(screen.getByText('Reference: #/components/responses/Success')).toBeInTheDocument();
    });
    it('handles empty operation object', () => {
        const props = {...defaultProps, operation: {}};
        setup(<EndpointDetail {...props} />);
        expect(screen.getByTestId('badge')).toHaveTextContent(/GET/i);
        expect(screen.getByText('/users')).toBeInTheDocument();
        expect(screen.queryByRole('table')).not.toBeInTheDocument();
        expect(screen.queryByText('Request Body')).not.toBeInTheDocument();
        expect(screen.queryByText('Responses')).not.toBeInTheDocument();
    });
    it('handles unknown HTTP method', () => {
        const props = {...defaultProps, method: 'INVALID', operation: {summary: 'Test'}};
        setup(<EndpointDetail {...props} />);
        expect(screen.getByTestId('badge')).toHaveTextContent('INVALID');
    });
    it('handles request body and responses with no content types', () => {
        const props = {
            ...defaultProps,
            method: 'post',
            operation: {
                requestBody: {content: {}},
                responses: {'200': {description: 'Success', content: {}}},
            },
        };
        setup(<EndpointDetail {...props} />);

        expect(screen.queryByText('Request Body')).not.toBeInTheDocument();

        const badges = screen.getAllByTestId('badge');
        expect(badges).toHaveLength(2);
        expect(badges[0]).toHaveTextContent('post');
        expect(badges[1]).toHaveTextContent('200');

        expect(screen.getByText('Success')).toBeInTheDocument();
        expect(screen.queryByTestId('schema-table')).not.toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const props = {
            ...defaultProps,
            operation: {
                responses: {
                    '200': {
                        description: 'Successful response',
                        content: {
                            'application/json': {schema: {type: 'array', items: {type: 'string'}}},
                        },
                    },
                },
            },
        };
        const {asFragment} = render(<EndpointDetail {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('handles large number of parameters efficiently', () => {
        const props = {
            ...defaultProps,
            operation: {
                parameters: Array(100)
                    .fill(null)
                    .map((_, i) => ({name: `param${i}`, in: 'query', schema: {type: 'string'}})),
            },
        };
        const start = performance.now();
        render(<EndpointDetail {...props} />);
        const end = performance.now();
        const table = screen.getByRole('table');
        const rows = within(table).getAllByRole('row');
        expect(rows).toHaveLength(101);
        expect(end - start).toBeLessThan(1000);
    });
});