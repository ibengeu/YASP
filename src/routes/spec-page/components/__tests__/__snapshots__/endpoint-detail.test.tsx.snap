// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EndpointDetail > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full overflow-hidden"
  >
    <div>
      <div
        class="p-6"
      >
        <div
          class="flex items-center gap-4 mb-4"
        >
          <span
            class="uppercase text-white bg-blue-500"
            data-testid="badge"
          >
            get
          </span>
          <code
            class="text-sm font-mono"
          >
            /users
          </code>
        </div>
        <div
          class="space-y-6"
        >
          <h4
            class="text-sm font-semibold mb-2"
          >
            Responses
          </h4>
          <div
            class="mb-6 border rounded-lg p-4 bg-stone-50/70"
          >
            <div
              class="flex items-center gap-2 mb-4"
            >
              <span
                class="text-sm"
                data-testid="badge"
                data-variant="default"
              >
                200
              </span>
              <span
                class="text-sm font-medium"
              >
                Successful response
              </span>
            </div>
            <div
              class="mt-2"
            >
              <span
                class="mb-2"
                data-testid="badge"
                data-variant="outline"
              >
                application/json
              </span>
              <div
                data-testid="schema-table"
              >
                {"type":"array","items":{"type":"string"}}
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
